cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -3557,4 +3602,240 @@ mod tests {
+                    r#"Group::"thing""#.parse().unwrap(),
+                ),
+                ActionConstraint::any(),
+                ResourceConstraint::any(),
+            ),
+            (
+                r#"permit(principal is A::User in Group::"thing", action, resource);"#,
+                PrincipalConstraint::is_type_in(
+                    "A::User".parse().unwrap(),
+                    r#"Group::"thing""#.parse().unwrap(),
+                ),
+                ActionConstraint::any(),
+                ResourceConstraint::any(),
+            ),
+            (
+                r#"permit(principal is User in ?principal, action, resource);"#,
+                PrincipalConstraint::is_type_in_slot("User".parse().unwrap()),
+                ActionConstraint::any(),
+                ResourceConstraint::any(),
+            ),
+            (
+                r#"permit(principal, action, resource is Folder);"#,
+                PrincipalConstraint::any(),
+                ActionConstraint::any(),
+                ResourceConstraint::is_type("Folder".parse().unwrap()),
+            ),
+            (
+                r#"permit(principal, action, resource is Folder in Folder::"inner");"#,
+                PrincipalConstraint::any(),
+                ActionConstraint::any(),
+                ResourceConstraint::is_type_in(
+                    "Folder".parse().unwrap(),
+                    r#"Folder::"inner""#.parse().unwrap(),
+                ),
+            ),
+            (
+                r#"permit(principal, action, resource is Folder in ?resource);"#,
+                PrincipalConstraint::any(),
+                ActionConstraint::any(),
+                ResourceConstraint::is_type_in_slot("Folder".parse().unwrap()),
+            ),
+        ] {
+            let policy = parse_policy_template(None, src).unwrap();
+            assert_eq!(policy.principal_constraint(), &p);
+            assert_eq!(policy.action_constraint(), &a);
+            assert_eq!(policy.resource_constraint(), &r);
+        }
+    }
+
+    #[test]
+    fn is_err() {
+        let invalid_is_policies = [
+            (
+                r#"permit(principal in Group::"friends" is User, action, resource);"#,
+                "expected a entity uid or template slot",
+            ),
+            (
+                r#"permit(principal, action, resource in Folder::"folder" is File);"#,
+                "expected a entity uid or template slot",
+            ),
+            (
+                r#"permit(principal is User == User::"Alice", action, resource);"#,
+                "`is` cannot appear in the scope at the same time as `==`",
+            ),
+            (
+                r#"permit(principal, action, resource is Doc == Doc::"a");"#,
+                "`is` cannot appear in the scope at the same time as `==`",
+            ),
+            (
+                r#"permit(principal is User::"alice", action, resource);"#,
+                r#"unexpected token `"alice"`"#,
```

Comment:
this error could maybe be better -- doesn't have to be for this PR but maybe file an issue to improve?  Same with some of the other errors around here

Reply:
Opened issue #409.